fix(get_scylla_docker_repo_from_version): revert #12970 and add custom dev ScyllaDB build logic

abramche opened the pull request:
- Add missing dev ScyllaDB build logic and unit tests to cover this logic.
- Revert #12970 as this change supersedes it (I cherry-picked the docstring from this PR though)
- Drop hardcoded k8s configuration from Jenkinsfile

### Testing
- [x] covered by existing and newly added unit tests

### PR pre-checks (self review)
- [x] I added the relevant `backport` labels
- [x] I didn't leave commented-out/debugging code

abramche commented on sdcm/sct_config.py:
```diff
@@ -3036,7 +3036,7 @@ def __init__(self):  # noqa: PLR0912, PLR0914, PLR0915
 
         if scylla_version := self.get("scylla_version"):
             if self.get("cluster_backend") in ["docker", "k8s-eks", "k8s-gke"]:
-                self["docker_image"] = get_scylla_docker_repo_from_version(scylla_version, self.get("docker_image"))
```
I do not understand a use case when we intentionally want to force a repository over the standard ones. As far as I understand, this override was re-introduced by Israel to spot fix dev build resolution (this is covered by the function instead now). We do not have custom repos for dev builds, do we?

abramche replied:
> You might want to test a specific image (e.g. your own image in your own repository) and for that the override is the correct way to do it.
> 
> It should not be a common usecase, but I do not think it needs to be removed, given how easy it is to support it.

got it - then I think the best way is not to touch the repo resolution function at all (as it does not "resolve" version but returns a hardcoded stub) but rather make the use case skip it altogether

```
            if self.get("cluster_backend") in ["docker", "k8s-eks", "k8s-gke"] and not self["docker_image"]:
```
and add SCTConfiguration unit test(s) to check this specific use case. WDYT?

pehala replied:
If that will work, I do not have problems with it

fruch commented on unit_tests/test_version_utils.py:
```diff
@@ -761,30 +761,33 @@ def test_get_branched_repo(scylla_version, distro, expected_repo):
+        ("branch-2024.1:latest", "scylladb/scylla-enterprise"),
+        ("2024.5.0-dev-0.20251217.55f4a2b75472", "scylladb/scylla-enterprise-nightly"),
+        ("2024.99.99-weirdnonsemver", "scylladb/scylla-enterprise-nightly"),
+        ("2025.0.0-dev-0.20251217.55f4a2b75472", "scylladb/scylla-enterprise-nightly"),
```
This is wrong any 2025 and up shouldn't be from enterprise

fruch replied:
And kind of pointless to test non valid versions, i.e. this won't exist in any repo...

vponomaryov commented:
> > The one failing unit test has nothing to do with these changes, 2025.4 got released just now (timing is hilarious though)
> > @fruch when you have a minute, mind double checking after possible version shenanigans? The original culprit is handled now (with unit tests too) - any more possible variations of the build format?
> > I am still unsure though if `master:latest` should pull from `scylladb/scylla` and not from `scylladb/scylla-nightly`; we already have a [test](https://github.com/scylladb/scylla-cluster-tests/blob/e298929269e4a73a270ad453820c1325c2b04b07/unit_tests/test_config_get_version_based_on_conf.py#L45) which asserts against `-nightly` - @vponomaryov you [asked](https://github.com/scylladb/scylla-cluster-tests/pull/12953#discussion_r2620281216) to match `master:latest` to `scylladb/scylla` - should I change it?
> 
> master:latest is a tricky one, since all dev are built into nightly repo, also the branches
> 
> `scylladb/scylla:latest` is latest official release, not master. `scylladb/scylla-nightly:latest` I think should be pointing only to master, and branches shouldn't update latest, but that need to be cross checked with releng.

Probably I was wrong and referring of the `master:latest` to the `scylladb/scylla-nightly:latest` really sounds proper.
The `scylladb/scylla-nightly:latest` is really about the latest build from the scylla master branch whereas the `scylladb/scylla:latest` refers to ANY latest build, even if it is a patch version to some old scylla like 2025.1...

> I don't understand in our code how branch-2025.1:latest work, since it need to look it up...

https://github.com/scylladb/scylla-cluster-tests/blob/b8bf758ccb1503ece2f99dac1a44e61f7678c984/sdcm/sct_config.py#L3050-L3079

abramche commented:
`master:latest` vs `latest` sure introduces a lot of confusion

I think we have not had SCT logic yet in place to compare which is the more recent, `scylladb/scylla-nightly:latest` or `scylladb/scylla:latest`

tbh it feels we overthink here. From a test POV you are more likely to request `latest` if you want the latest possible nightly build.
If you need the latest released build, you might as well just override with `docker_image = 'scylladb/scylla'` in the test logic.
In short, `master:latest` and `latest` may both just point to `-nightly:latest` then if you rely on automatic repo resolution, WDYT?

abramche commented:
was on PTO, picking it up again

> and for `branch-2025.4:latest` let remove it from testing, it's need further lookups to really work, since we don't have a repo per branch

do you mean the tests only or the logic too? It feels wrong to remove it from the tests and leave the (untested) lookup logic, even more so when the tests are so cheap. 

I addressed the rest of the comments now

abramche commented:
> when it was written it wasn't clear where 2025.1 would reside, and the test was written base on existing logic at the time. yes, it should change to assume `scylladb/scylla`

after reviewing this test I introduced the following changes:
- 2025.1 -> `scylladb/scylla` repo
- fixed [another logical problem](https://github.com/scylladb/scylla-cluster-tests/blob/5d0ce8f09af8a45c915e19ee8a0878205ba46e45/sdcm/sct_config.py#L4036) (`is_enterprise` being assumed based off scylla repository and not feature set of a version)
- set expected enterprise flag to `True` for both `latest` and `master:latest` - the returned builds should always have the enterprise feature set
- ruff was complaining about unused variable in `sct_config.py` so I dropped it; idk how it passed the ruff check before

The tests were still failing after that as `master:latest` and `enterprise` versions were not discovered as enterprise builds so
⚠️ I changed the `is_enterprise` function to return True on scylla_version in ('enterprise', 'master:latest', 'latest') - this is a small change with big impact - but it makes sense to me - it will show other invalid assumptions in SCT tests on failures

I'll push a temporary new commit to check how this change reflects on all tests since my local environment is currently unreliable

abramche commented on sdcm/utils/version_utils.py:
```diff
@@ -450,7 +453,9 @@ def is_enterprise(scylla_version):
     """
-    return bool(re.search(r"^20[0-9]{2}.*", scylla_version))
+    return bool(
+        re.search(r"^20[0-9]{2}.*", scylla_version) or scylla_version in ("enterprise", "master:latest", "latest")
```
> We don't test Scylla versions which must return `False` to the condition you fix.

[You](https://github.com/scylladb/scylla-cluster-tests/blob/e19852d3ce64e5582a27fe479bfb048b7b71fe93/unit_tests/test_config_get_version_based_on_conf.py#L44) [do](https://github.com/scylladb/scylla-cluster-tests/blob/e19852d3ce64e5582a27fe479bfb048b7b71fe93/unit_tests/test_config_get_version_based_on_conf.py#L45) - both these scenarios try to wrongly assume and test that `master:latest` and `latest` are open source. And these tests [fail](https://github.com/scylladb/scylla-cluster-tests/pull/13084) in `master` and are fixed in this PR.

> So, if we do something with it now then it is removal of it making every place, which uses it, behave like we have `True` condition result.

If old version/OSS is out of scope now, sure, we can axe it. But it's a separate, independent change with a lot bigger scope than this PR.

*
Anyway, my changes do not break any test, as evident by comparing tests from master against tests from my PR; so what is holding them back?

abramche commented on sdcm/sct_config.py:
```diff
@@ -4031,9 +4031,8 @@ def get_version_based_on_conf(self):
         elif backend == "docker" or "k8s" in backend:
-            docker_repo = self.get("docker_image")
             scylla_version = self.get("scylla_version")
-            _is_enterprise = "enterprise" in docker_repo
```
resolving as these changes will be done in scope of https://github.com/scylladb/scylla-cluster-tests/issues/13093 and are not a part of this PR any more

abramche commented:
> LGTM
> 
> But need to fix one more test that failed in the integration tests Ci job:
> 
> ```
> 21:52:58  FAILED unit_tests/test_config_get_version_based_on_conf.py::test_docker[2025.1] - AssertionError: assert 'scylladb/scylla' == 'scylladb/scylla-enterprise'
> 21:52:58    
> 21:52:58    - scylladb/scylla-enterprise
> 21:52:58    + scylladb/scylla
> ```

No. This test has nothing to do with these changes and [already exists in master](https://github.com/scylladb/scylla-cluster-tests/pull/13102). The reason it is failing - it is a bad test that needs to be rewritten.

I already made a good effort to fix this in scope of this PR; since you rejected my fixes, it will have to be done in scope of https://github.com/scylladb/scylla-cluster-tests/issues/13093 by someone else (or me, if/when I have free time)

dimakr commented:
> @dimakr @soyacz ^ care to take a look, please?

Nothing to do with the test. There is a [known issue](https://github.com/scylladb/vector-store/pull/330) in vector store itself, fix is in progress

vponomaryov reviewed (approved):
LGTM

@fruch since integration tests have "fail" status this PR must be force-merged, I don't have such permissions
